ledger-service/fetch-contracts/src/main/scala/fetchcontracts/domain.scala

```diff
@@ -96,24 +99,44 @@ package domain {
       a.key.fold(false)(_ == k)
 
     def fromLedgerApi(
-        gacr: lav1.active_contracts_service.GetActiveContractsResponse
+        resolvedQuery: domain.ResolvedQuery,
+        gacr: lav1.active_contracts_service.GetActiveContractsResponse,
     ): Error \/ List[ActiveContract[lav1.value.Value]] = {
-      gacr.activeContracts.toList.traverse(fromLedgerApi(_))
+      gacr.activeContracts.toList.traverse(fromLedgerApi(resolvedQuery, _))
     }
 
-    def fromLedgerApi(in: lav1.event.CreatedEvent): Error \/ ActiveContract[lav1.value.Value] =
+    def fromLedgerApi(
+        resolvedQuery: domain.ResolvedQuery,
+        in: lav1.event.CreatedEvent,
+    ): Error \/ ActiveContract[lav1.value.Value] = {
+
+      def getIdAndPayload: (Error\/ lav1.value.Identifier, Error\/ lav1.value.Record) = resolvedQuery match {
+        case ResolvedQuery.ByInterfaceId(interfaceId) =>
+          // TODO is there an easier way to do this?
+          val id = lav1.value.Identifier(packageId = interfaceId.packageId, moduleName = interfaceId.moduleName, entityName = interfaceId.entityName)
+          val payload = in
+            .interfaceViews
+            .find(_.interfaceId.exists(_ == id))
+            .flatMap(_.viewValue) required "interfaceView"
+          (\/-(id),payload)
+        case _ =>
+          (in.templateId required "templateId", in.createArguments required "createArguments")
+      }
+
+      val (getId, getPayload) = getIdAndPayload
       for {
-        templateId <- in.templateId required "templateId"
-        payload <- in.createArguments required "createArguments"
+        id <- getId
+        payload <- getPayload
       } yield ActiveContract(
         contractId = ContractId(in.contractId),
-        templateId = ContractTypeId.Template fromLedgerApi templateId,
+        templateId = ContractTypeId.Template fromLedgerApi id,
```

Comment:
We want a `ContractTypeId.Interface` to be used in the interface case. This can be added to your `match` `case`s and tuple above.

Reply:
Almost missed this comment should be good now.